### ISSUE TITLE
Versioning

### ISSUE THREAD

**Issue by fimbault** [vuejs/vitepress#109]
**Is your feature request related to a problem? Please describe.**
Related to documentation versioning

**Describe the solution you'd like**
A way to version a set of markdown documents

**Describe alternatives you've considered**
Similar to [VuePress Plugin](https://www.npmjs.com/package/vuepress-plugin-versioning) or [Docusaurus2](https://v2.docusaurus.io/docs/versioning)

**Comment by brc-dd** [comment]
Well there is another workaround. Move your old docs to a subdomain. Then you can write config like this:

```js
import { defineConfig } from 'vitepress'

export default defineConfig({
  themeConfig: {
    nav: [
      {
        text: 'v3 (stable)',
        items: [
          {
            text: 'v2 (legacy)',
            link: 'https://v2.vuejs.org/'
          }
        ]
      }
    ]
  }
})
```

It will be rendered like this:

![image](https://user-images.githubusercontent.com/40380293/171828967-b6547c51-d72c-4137-8fae-14b2be0a1c33.png)

**Comment by jcstein** [comment]
hi again @brc-dd - do you have recommended practices on moving old versions to subdomain? if you do, have you seen anyone do this on github pages? we're thinking of moving another docs site to vitepress, but this is the one remaining blocker

**Comment by DMiradakis** [comment]
+1 this would really be essential. I'm trying to think of how to build this in VitePress, but I'm not sure at the moment.

**Comment by drwpow** [comment]
I’m pretty sure this is just done with [multiple sidebars](https://vitepress.dev/reference/default-theme-sidebar#multiple-sidebars). Just move your versioned docs to a subfolder with the version in them.

You can then add a link to the [nav](https://vitepress.dev/reference/default-theme-nav) where a user selects the version, and on change, the correct sidebar will show.

**Comment by IMB11** [comment]
I've utilized multiple sidebars, a nav switcher etc. to create a versioning example that should suffice for most projects until proper versioning support is added. 

You can view it here: https://github.com/IMB11/vitepress-versioning

**Comment by ciscoo** [comment]
> Well there is another workaround. Move your old docs to a subdomain. Then you can write config like this:

I would not consider this a workaround primarily because it assumes you have a domain. For users that use GitHub Pages without a custom domain, it is not possible.

A workaround, in my opinion, is something that can be done purely within the constraints of VitePress. The approaches that @drwpow and @IMB11 provided I would consider workarounds.

**Comment by IMB11** [comment]
I actually ended up turned my example into a proper plugin/library: https://www.npmjs.com/package/vitepress-versioning-plugin

**Comment by ajiho** [comment]
> I actually ended up turned my example into a proper plugin/library: https://www.npmjs.com/package/vitepress-versioning-plugin

Thank you for your work, but I think we need Vitepress official to implement it better. Your package seems to have a different version control effect from the one provided by docusaurus

**Comment by IMB11** [comment]
Well, working without an official plugin API is extremely difficult - there's some whacky stuff going on in that plugin that would be 10x easier if there was a standardized approach to plugins.

**Comment by ajiho** [comment]
@brc-dd Will version control functionality be introduced in the future? I have tried the version control feature of docusaurus and I think it would be a satisfactory answer. Can the Vitepress team migrate it?

**Comment by javierjulio** [comment]
I've been migrating ActiveAdmin's documentation to Vitepress. The setup, experience and the overall docs are just fantastic, thank you! ❤️ The one issue I've run into is we do need the capability or an alternative for versioning without requiring a subdomain. If Vitepress could support some way of doing documentation versioning, that would be great and helpful feature.

**Comment by aminnairi** [comment]
First of all, thank you for making Vitepress, very nice to work with.

I think at this point, we can all agree to say that Vitepress is mainly used for tech docs websites, and that a first-class support for versioning is expected from such a tool.

**Comment by brc-dd** [comment]
In v2-alpha, overriding themeConfig per directory is supported. So you can have things like this - https://stackblitz.com/edit/vite-1lzrkv7b

Fully polished versioning support still needs us to implement some other things including:
- Version switcher component (navbar currently supports custom components, but it will be better if this is built-in, and it should keep url other than the version same while switching, similar to how current locale switcher works)
- Redirects to latest version (although this should probably be configured at hosting provider level)
- Search faceting
- Unversioned pages (this is gonna be bit tricky to implement, probably will need to implement falling back to root if missing behavior in locales, and then adapt that implementation so that you can store unversioned pages only in the latest version)

Please share if you all have any other feature requests for versioning support.

**Comment by hamishwillee** [comment]
Yes, versions, and ideally translations, should be decoupled in the build. Specifically, I should be able to build a version and or a translation at different times, put them in the right folder structure for my config, and still have them link.

My docs are currently about 700 pages, with 3 translations. That takes about 7 minutes to build. If I add a version I'd expect that to double, which is not scalable. 

In generally it would be better if I can build the parts separately, because sometimes the translations fail due to round trip errors in the translation service - it would be good if that didn't break the rest of my build. Hope this makes sense.

**Comment by axsaucedo** [comment]
+1

**Comment by dynamsoft-h** [comment]
Does the team plan on shipping this feature for 2.0, and will it support git versioning with branches or tags? Vitepress has been a joy to use thus far.
